planner/core/hashcode.go

```diff
@@ -97,3 +76,266 @@ func (p *LogicalLimit) HashCode() []byte {
+func (p *LogicalUnionAll) HashCode() []byte {
+	// PlanType + SelectOffset
+	result := make([]byte, 0, 8)
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalApply) HashCode() []byte {
+	// p.LogicalJoin.HashCode() + Encode(CorCols)
+	result := p.LogicalJoin.HashCode()
+	corColHashCode := func(i int) []byte { return p.CorCols[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.Encode(result, corColHashCode, len(p.CorCols))
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalWindow) HashCode() []byte {
+	result := make([]byte, 0, 48+len(p.WindowFuncDescs)*29+14*(len(p.PartitionBy)+len(p.OrderBy)))
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+
+	wfdHashCode := func(i int) []byte { return p.WindowFuncDescs[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.Encode(result, wfdHashCode, len(p.WindowFuncDescs))
+
+	partitionByHashCode := func(i int) []byte { return p.PartitionBy[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.EncodeAndSort(result, partitionByHashCode, len(p.PartitionBy))
+
+	orderByHashCode := func(i int) []byte { return p.OrderBy[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.Encode(result, orderByHashCode, len(p.OrderBy))
+
+	if p.Frame != nil {
+		result = append(result, p.Frame.HashCode(p.ctx.GetSessionVars().StmtCtx)...)
+	}
+
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalMemTable) HashCode() []byte {
+	result := make([]byte, 0, 37)
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+
+	result = codec.EncodeInt(result, p.TableInfo.ID)
+	result = codec.EncodeCompactBytes(result, hack.Slice(p.DBName.L))
+
+	result = codec.EncodeInt(result, p.QueryTimeRange.From.Unix())
+	result = codec.EncodeInt(result, p.QueryTimeRange.To.Unix())
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalUnionScan) HashCode() []byte {
+	result := make([]byte, 0, 30+13*(len(p.conditions)))
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+	result = append(result, p.handleCol.HashCode(p.ctx.GetSessionVars().StmtCtx)...)
+	condHashCode := func(i int) []byte { return p.conditions[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.EncodeAndSort(result, condHashCode, len(p.conditions))
+	result = append(result, p.handleCol.HashCode(p.ctx.GetSessionVars().StmtCtx)...)
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *TiKVSingleGather) HashCode() []byte {
+	dsHashCode := p.Source.HashCode()
```

Comment:
We don't need the whole hashcode of the datasource. Only the table identity of the datasource is enough.